plutus-core/plutus-core/src/PlutusCore/Constant/Meaning.hs

```diff
@@ -232,94 +230,73 @@ type x ~?~ y = TryUnify (x === y) x y
 -- | Get the element at an @i@th position in a list.
 type Lookup :: forall a. Nat -> [a] -> a
 type family Lookup n xs where
+    Lookup _ '[]       = TypeError ('Text "Not enough elements")
     Lookup 0 (x ': xs) = x
     Lookup n (_ ': xs) = Lookup (n - 1) xs
 
 -- | Get the name at the @i@th position in the list of default names. We could use @a_0@, @a_1@,
 -- @a_2@ etc instead, but @a@, @b@, @c@ etc are nicer.
-type GetName i = Lookup i '["a", "b", "c", "d", "e", "f", "g", "h"]
+type GetName :: GHC.Type -> Nat -> Symbol
+type family GetName k i where
+    GetName GHC.Type i = Lookup i '["a", "b", "c", "d", "e", "i", "j", "k", "l"]
+    GetName _        i = Lookup i '["f", "g", "h", "m", "n"]
+
+-- | Like 'id', but a type constructor.
+type Id :: forall a. a -> a
+data family Id x
 
 -- | Try to specialize @a@ as a type representing a PLC type variable.
 -- @i@ is a fresh id and @j@ is a final one (either @i + 1@ or @i@ depending on whether
 -- specialization attempt is successful or not).
-type TrySpecializeAsVar :: Nat -> Nat -> GHC.Type -> GHC.Type -> GHC.Constraint
-class TrySpecializeAsVar i j term a | i term a -> j
+-- @f@ is for wrapping 'TyVarRep' (see 'HandleHole' for how this is used).
+type TrySpecializeAsVar :: forall k. Nat -> Nat -> (k -> k) -> k -> GHC.Constraint
+class TrySpecializeAsVar i j f a | i f a -> j
 instance
-    ( var ~ Opaque term (TyVarRep ('TyNameRep (GetName i) i))
+    ( var ~ f (TyVarRep @k ('TyNameRep (GetName k i) i))
     -- Try to unify @a@ with a freshly created @var@.
     , a ~?~ var
     -- If @a@ is equal to @var@ then unification was successful and we just used the fresh id and
     -- so we need to bump it up. Otherwise @var@ was discarded and so the fresh id is still fresh.
     -- Replacing @(===)@ with @(==)@ causes errors at use site, for whatever reason.
     , j ~ If (a === var) (i + 1) i
-    ) => TrySpecializeAsVar i j term a
-
--- | For looking under special-case types, for example the type of a constant or the type arguments
```

Comment:
`master`:
![Screenshot from 2022-01-07 22-42-36](https://user-images.githubusercontent.com/10480926/148598382-d6520e33-9ffd-4a05-a0cb-042f70e9f3df.png)

this branch:
![Screenshot from 2022-01-07 22-43-04](https://user-images.githubusercontent.com/10480926/148598353-fde925f7-2d46-4cf8-8b36-206ae1ccb310.png)